File: 442.find-all-duplicates-in-an-array.ts
/*
 * @lc app=leetcode id=442 lang=typescript
 *
 * [442] Find All Duplicates in an Array
 */

// @lc code=start
function findDuplicates(nums: number[]): number[] {
  const duplicates: { [key: number]: number } = {};

  for (let num of nums) {
    if (duplicates[num]) {
      duplicates[num]++;
    } else {
      duplicates[num] = 1;
    }
  }

  return Object.keys(duplicates)
    .filter((key) => duplicates[key] === 2)
    .map(Number);
}
// @lc code=end

File: go/442.find-all-duplicates-in-an-array.go
func findDuplicates(nums []int) []int {
	results := make([]int, 0)

	for i := 0; i < len(nums); i++ {
		num := abs(nums[i])
		idx := num - 1
		if nums[idx] < 0 {
			results = append(results, num)
		}
		nums[idx] *= -1
	}

	return results
}

func abs(num int) int {
	if num < 0 {
		return num * -1
	}

	return num
}